Formula/mingw-w64.rb

```diff
@@ -0,0 +1,177 @@
+        --with-isl=#{Formula["isl"].opt_prefix}
+        MAKEINFO=#{Formula["texinfo"].opt_bin}/makeinfo
+      ]
+      if target_arch.start_with?("i686")
+        gcc_args << "--disable-multilib"
+      elsif target_arch.start_with?("x86_64") && build.without?("multilib")
+        gcc_args << "--enable-64bit" << "--disable-multilib"
+      elsif target_arch.start_with?("x86_64") && build.with?("multilib")
+        gcc_args << "--enable-targets=all"
+      end
+
+      mkdir "#{buildpath}/gcc/build-#{target_arch}" do
+        system "../configure", *gcc_args
+        system "make", "all-gcc"
+        system "make", "check-gcc"
+        system "make", "install-gcc"
+      end
+
+      ENV.prepend_path "PATH", bin.to_s
+
+      args = %W[
+        CC=#{target_arch}-gcc
+        CXX=#{target_arch}-g++
+        CPP=#{target_arch}-cpp
+        LD=#{target_arch}-gcc
+        --host=#{target_arch}
+        --prefix=#{prefix}/#{target_arch}
+        --with-sysroot=#{prefix}/#{target_arch}
+      ]
+
+      if target_arch.start_with?("i686")
+        args << "--enable-lib32" << "--disable-lib64"
+      elsif target_arch.start_with?("x86_64") && build.without?("multilib")
+        args << "--disable-lib32" << "--enable-lib64"
+      elsif target_arch.start_with?("x86_64") && build.with?("multilib")
+        args << "--enable-lib32" << "--enable-lib64"
+      end
+
+      mkdir "mingw-w64-crt/build-#{target_arch}" do
+        system "../configure", *args
+        system "make"
+        system "make", "install"
+      end
+
+      chdir "#{buildpath}/gcc/build-#{target_arch}" do
+        system "make"
+        system "make", "install"
+      end
+
+      ln_s "../../lib/gcc/#{target_arch}/lib/libgcc_s.a", "#{prefix}/#{target_arch}/lib"
+
+      ENV["LDPATH"] = "#{target_arch}/#{lib}"
+      args = %W[
+        CC=#{target_arch}-gcc
+        CXX=#{target_arch}-g++
+        CPP=#{target_arch}-cpp
+        --host=#{target_arch}
+        --prefix=#{prefix}/#{target_arch}
+      ]
+
+      mkdir "mingw-w64-libraries/winpthreads/build-#{target_arch}" do
+        system "../configure", *args
+        system "make"
+        system "make", "install-strip"
+      end
+    end
+  end
+
+  test do
+    (testpath/"hello.c").write <<-EOS.undent
+      #include <stdio.h>
+      int main()
+      {
+      puts("Hello, world!");
+      return 0;
+      }
+      EOS
+    (testpath/"hello.cc").write <<-EOS.undent
+      #include <iostream>
+      int main()
+      {
+      std::cout << "Hello, world!" << std::endl;
+      return 0;
+      }
+      EOS
+    (testpath/"hello.f90").write <<-EOS.undent
+      program hello
+      print *, "Hello, world!"
+      end program hello
+      EOS
+    target_archs.each do |target_arch|
+      system "#{bin}/#{target_arch}-gcc", "-o", "hello-c.exe", "hello.c"
+      system "#{bin}/#{target_arch}-g++", "-o", "hello-cc.exe", "hello.cc"
+      system "#{bin}/#{target_arch}-gfortran", "-o", "hello-f90.exe", "hello.f90"
+    end
```

Comment:
Can we test, at least slightly, that the executables have the right format?

```
assert_match "PE32+ executable (console) x86-64, for MS Windows", shell_output("file hello-f90.exe")
assert_match "file format pei-x86-64", shell_output("#{bin}/x86_64-w64-mingw32-objdump -a hello-f90.exe")
```

(that's for 64-bit, but we can adapt the patterns to match each target)